### PR TITLE
REF: Add `Extract trait` refactoring

### PR DESCRIPTION
**Pull request by dima74**
output

Part of #8326

changelog: Add `Extract trait` refactoring

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+import org.rust.openapiext.isUnitTestMode
+import javax.swing.JComponent
+import javax.swing.JTextField
+
+class RsExtractTraitHandler : RefactoringActionHandler {
+    override fun invoke(project: Project, editor: Editor, file: PsiFile, dataContext: DataContext?) {
+        val offset = editor.caretModel.offset
+        val element = file.findElementAt(offset) ?: return
+        val impl = element.ancestorOrSelf<RsImplItem>() ?: return
+        if (impl.traitRef != null) return
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt:43]: Do we have any plans to support `impl Foo for Bar` impl blocks eventually? If so, we could add a comment here. Also, probably would be great to open an issue to track the progress on `Extract trait` refactoring in general

File: src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        if (!CommonRefactoringUtil.checkReadOnlyStatus(project, impl)) return
+
+        val members = (impl.members ?: return).childrenOfType<RsItemElement>()
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt:46]: I'm wondering if [inner attributes](https://doc.rust-lang.org/reference/items/implementations.html?highlight=impl#attributes-on-implementations) are covered here, could you confirm it or add a comment (and maybe a test)?

File: src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    }
+}
+
+class RsExtractTraitDialog(
+    project: Project,
+    private val impl: RsImplItem,
+    private val memberInfos: List<RsMemberInfo>,
+) : RefactoringDialog(project, false) {
+
+    private var traitNameField: JTextField = JBTextField().apply {
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt:69]: Is `var` necessary here?

File: src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+    override fun validateButtons() {
+        super.validateButtons()
+        previewAction.isEnabled = false
+    }
+
+    override fun areButtonsValid(): Boolean =
+        isValidRustVariableIdentifier(traitNameField.text) && memberInfos.any { it.isChecked }
+
+    @VisibleForTesting
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt:102]: I'm not sure if this annotation makes sense here. Looks like this method is public and also used deep inside the platform (outside the tests), WDYT?

File: src/test/kotlin/org/rust/ide/refactoring/RsExtractTraitTest.kt [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        impl<A> Trait for S<A> {
+            fn b(&self) {}
+        }
+
+        trait Trait {
+            fn b(&self);
+        }
+    """)
+
+    fun `test generics copmlex`() = doTest("""
```
Comment (ortem) [inline src/test/kotlin/org/rust/ide/refactoring/RsExtractTraitTest.kt:162]:
```suggestion
    fun `test generics complex`() = doTest("""
```

**Review by ortem — CHANGES_REQUESTED** [review]
LGTM in general except for minor comments. I also have a couple of questions:

1. Have you tried renaming the corresponding refactoring from `Extract interface` to `Extract trait` somehow? I'm not sure if it's possible but I found similar [implementation for Scala](https://github.com/JetBrains/intellij-scala/blob/bce780a19af9ac6ff42522c625d6dac1a52cecd7/scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/ScalaRefactoringSupportProvider.scala) and its [registration](https://github.com/JetBrains/intellij-scala/blob/5014305b193838635d8ac207bd32d68a96eb1960/scala/scala-impl/resources/META-INF/scala-plugin-common.xml#L1957) that doesn't use `RefactoringSupportProvider` (not sure if it's OK for us though).
2. Could we change the behaviour to revoke all the changes by a single Undo action? Currently, seems like it takes two Undo actions -- one to delete the new trait and another one to restore the original.

File: src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt (hunk 1 above)
Comment (dima74) [reply src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt:43]: It is not very clear how `Extract trait` refactoring should work for trait impls, so no such plans at the moment. I plan to support the refactoring for `impl Struct` (this PR) and `impl Trait` (next PR)

File: src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt (hunk 2 above)
Comment (dima74) [reply src/main/kotlin/org/rust/ide/refactoring/RsExtractTrait.kt]: Thanks, attributes were not supported, I fixed and added the test

File: src/main/kotlin/org/rust/ide/refactoring/extractTrait/RsExtractTraitHandler.kt [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    private val memberInfos: List<RsMemberInfo>,
+) : RefactoringDialog(project, false) {
+
+    private val traitNameField: JTextField = JBTextField().apply {
+        addTextChangeListener { validateButtons() }
+    }
+
+    init {
+        super.init()
+        title = "Extract Trait"
```
Comment (ortem) [inline src/main/kotlin/org/rust/ide/refactoring/extractTrait/RsExtractTraitHandler.kt:73]: I found several `"Extract Trait"` strings in the code. Maybe reuse `action.Rust.RsExtractTrait.text` instead?

File: src/main/resources/messages/RsBundle.properties [hunk 7]
```diff
@@ -53,6 +53,8 @@ action.Rust.RsExtractStructFields.choose.fields.title=Choose Fields to Extract
 action.Rust.RsExtractStructFields.choose.name.dialog.title=Enter Name For the New Struct
 action.Rust.RsExtractStructFields.choose.name.dialog.invalid.name=Invalid struct name
 action.Rust.RsExtractStructFields.command.name=Extracting Struct Fields
+action.Rust.RsExtractTrait.text=Extract Trait
```
Comment (ortem) [inline src/main/resources/messages/RsBundle.properties:56]: Maybe use ellipsis `…` here, as in other refactoring names?
Screenshot 2021-12-21 at 17 58 39

File: src/test/kotlin/org/rust/ide/refactoring/RsExtractTraitTest.kt [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+
+package org.rust.ide.refactoring
+
+import org.intellij.lang.annotations.Language
+import org.rust.RsTestBase
+import org.rust.ide.refactoring.extractTrait.RS_EXTRACT_TRAIT_MEMBER_IS_SELECTED
+import org.rust.lang.core.psi.ext.RsItemElement
+import org.rust.lang.core.psi.ext.ancestorOrSelf
+
+class RsExtractTraitTest : RsTestBase() {
```
Comment (ortem) [inline src/test/kotlin/org/rust/ide/refactoring/RsExtractTraitTest.kt:14]: Let's add some tests with (doc) comments. Currently, the comments are always copied as I see. Later we could add an option to move/copy/ignore it, like Java or Kotlin implementations do with JavaDoc or KDoc. We could mention it in the tracking issue for this refactoring

**Review by ortem — CHANGES_REQUESTED** [review]
Looks even better now, great job! I have just a few comments here

**Review by ortem — APPROVED** [review]
bors r+

File: src/test/kotlin/org/rust/ide/refactoring/RsExtractTraitTest.kt (hunk 8 above)
Comment (ortem) [reply src/test/kotlin/org/rust/ide/refactoring/RsExtractTraitTest.kt:14]: BTW I've opened a tracking issue, feel free to add some further steps there
https://github.com/intellij-rust/intellij-rust/issues/8326

**Comment by ortem** [comment]
For QA: please check that Undo action works properly for this refactoring (single Undo should restore everything). It didn't work for me at some point ([comment](https://github.com/intellij-rust/intellij-rust/pull/8132#pullrequestreview-836860523)), but now I cannot reproduce it anymore, so hopefully, it got fixed.

**Comment by MarinaKalashina** [comment]
Docs done https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-refactorings.html#extract-trait

**Pull request merged**